### PR TITLE
ng_ndp: cache determined next-hop in FIB

### PR DESCRIPTION
**Pull request by authmillenon**
[RFC 4861](https://tools.ietf.org/html/rfc4861#section-5.2) recommends the store the determined next-hop in the "Destination Cache". For my implementation I use the FIB as such, if it is availabe, so I cache the next hops not fount in the FIB also in there to shorten the determination for later packages.

### CODE REVIEW

File: sys/net/network_layer/ng_ndp/ng_ndp.c [hunk 1]
```diff
@@ -370,11 +370,25 @@ kernel_pid_t ng_ndp_next_hop_l2addr(uint8_t *l2addr, uint8_t *l2addr_len,
             (ng_ipv6_netif_addr_get(prefix)->flags &
              NG_IPV6_NETIF_ADDR_FLAGS_NDP_ON_LINK)) {
             next_hop_ip = dst;
+#ifdef MODULE_FIB
+            /* We don't care if FIB is full, this is just for efficiency
+             * for later sends */
+            fib_add_entry(iface, (uint8_t *)dst, sizeof(ng_ipv6_addr_t), 0,
+                          (uint8_t *)next_hop_ip, sizeof(ng_ipv6_addr_t), 0,
+                          UINT32_MAX);
```
Comment (BytesGalore) [inline sys/net/network_layer/ng_ndp/ng_ndp.c]: please change to `FIB_LIFETIME_NO_EXPIRE` (same below)

**Comment by BytesGalore** [comment]
The added FIB entries need to be removed if the neighbour node is not reachable anymore, since then a next-hop request (if the FIB is active) will always successfully lead to nirvana. 
Probably this can be done here [1]?

[1] https://github.com/authmillenon/RIOT/blob/ng_ndp/enh/cache-next-hop/sys/net/network_layer/ng_ndp/ng_ndp.c#L262

**Comment by authmillenon** [comment]
@BytesGalore does `fib_remove_entry()` remove the entry to, if `dst` is the `next_hop` entry? For l376ff it isn't that problematic if it doesn't, but for 380ff I insert the default router as next hop for `dst`, not `dst` itself. Alternative could be to use the default router lifetime there as lifetime.

**Comment by authmillenon** [comment]
Also, can routing protocols cope with the fact, that this would "randomly" remove stuff from the FIB. @Lotterleben @cgundogan?

**Comment by BytesGalore** [comment]
> @BytesGalore does fib_remove_entry() remove the entry too, if dst is the next_hop entry?

yup, no mercy. The FIB do not determine if an entry is allowed to be deleted.

@authmillenon you're right this is a problem since usually each routing protocol is responsible for a specific prefix.
Changing entries handled by them will result in topology and routing inconsistencies.

**Comment by authmillenon** [comment]
So maybe I should give them a "reasonable lifetime" instead?

I think about
* `next_hop == ip`: reachability time, and
* `next_hop == default_router`: router lifetime

**Comment by BytesGalore** [comment]
+1 for the reasonable lifetimes if they expire the FIB entry is removed automatically.
> Also, can routing protocols cope with the fact, that this would "randomly" remove stuff from the FIB. 

The FIB can handle access and manipulation from different sources.
If the destination addresses have no collisions then everything is fine :)

**Comment by authmillenon** [comment]
Since the routing protocols probably use prefixes for the next-hop entries and I am using the full addresses, there should be no collision right?

Just one OT question popping-up: How ARE prefixes handled in the FIB? Up until now I assumed this is the `dst_len` of the add function, but this is in byte, right? Prefixes are usually given in bits (and with IPv4 the thing gets even more complicated with subnet-masks....)

**Comment by BytesGalore** [comment]
> Since the routing protocols probably use prefixes for the next-hop entries and I am using the full addresses, there should be no collision right?

If this is the case right :)
But I'm not certain if the routing protocols only use prefixes for next-hops.

> How ARE prefixes handled in the FIB?

Currently with bytes (I know this is wrong :blush:) but,
when #2818 is merged the FIB uses Bits for a prefix. Additionally with #3157 a prefix entry in the FIB is marked with a specific flag.

**Comment by BytesGalore** [comment]
oh, and when using IPv4, the subnet mask shouldn't be needed when an entry is flagged as prefix, since the trailing 0-Bits define the size of the subnet.

**Comment by authmillenon** [comment]
> oh, and when using IPv4, the subnet mask shouldn't be needed when an entry is flagged as prefix, since the trailing 0-Bits define the size of the subnet.

But weren't subnet masks introduced because there weren't enough prefixes left, making subnets defined by stuff like e.g. 63.255.0.0 possible, too.

**Comment by authmillenon** [comment]
(I know this discussion is going a little bit meta now :D will fix the timeouts soon).

**Comment by authmillenon** [comment]
oh and regarding overwriting entries of RPs: this should be more or less impossible, since in l351 it is checked if there is already something in the FIB. I say "more or less" since I think the IPv6 thread can't be preempted by the routing protocol in this function, but I might be wrong there.

**Comment by authmillenon** [comment]
To recap: I assume that it is safe to just remove the destination if the neighbor becomes unreachable, since it is unlikely that the routing protocol will add it exactly as in this function.

**Comment by BytesGalore** [comment]
> But weren't subnet masks introduced because there weren't enough prefixes left, 

somehow. If I remember right they were extended form Class-Network prefixes only to enable an arbitrary number of significant bits for private networks.

> making subnets defined by stuff like e.g. 63.255.0.0 possible, too?

I can't tell for sure, but feels like not (or a proprietary and error-prone solution).

**Comment by authmillenon** [comment]
Rebased and had to squash, since `ng_ndp.c` was too different now.

**Comment by BytesGalore** [comment]
ACK, lets see what travis says

**Pull request merged**